src/UI.ts

```diff
@@ -0,0 +1,103 @@
+import {
+  MessageActionRow,
+  MessageButtonOptions,
+  MessageButtonStyleResolvable,
+} from 'discord.js';
+import { ButtonHandler } from './ButtonHandler';
+import { v4 as uuidv4 } from 'uuid';
+
+export type UIComponent = DispatchButton | DispatchLinkButton;
+
+export type ButtonOptions = Omit<
+  MessageButtonOptions,
+  'customId' | 'style' | 'url'
+> & {
+  style: Exclude<MessageButtonStyleResolvable, 'LINK'>;
+  onClick: ButtonHandler;
+};
+
+export type LinkButtonOptions = Omit<
+  MessageButtonOptions,
+  'customId' | 'style' | 'url'
+> & {
+  url: string;
+};
+
+export class DispatchButton {
+  constructor(readonly options: ButtonOptions) {}
+
+  toDiscordComponent(
+    buttonListeners: Map<string, ButtonHandler>
+  ): MessageButtonOptions {
+    const { onClick, ...options } = this.options;
+    // nonlink buttons must have a customId
+    const id = getID(this.options.label ?? '<unlabeled>', 'button');
+    buttonListeners.set(id, onClick);
+    return { ...options, type: 'BUTTON', customId: id };
+  }
+}
+
+export class DispatchLinkButton {
+  constructor(readonly options: LinkButtonOptions) {}
+
+  toDiscordComponent(): MessageButtonOptions {
+    // we override style in case the user omitted it
+    return {
+      ...this.options,
+      type: 'BUTTON',
+      style: 'LINK',
+    };
+  }
+}
```

Comment:
I thought about doing that. Before I thought this was a better solution because it makes the type errors better when the user doesn't include the correct properties.